src/Core/Checkout/Promotion/Cart/PromotionCalculator.php

```diff
@@ -368,19 +380,105 @@ private function limitDiscountResult(float $maxDiscountValue, PriceCollection $p
      */
     private function isRequirementValid(LineItem $lineItem, Cart $calculated, SalesChannelContext $context): bool
     {
-        // if we dont have any requirement, then it's obviously valid
-        if (!$lineItem->getRequirement()) {
+        $requirement = $lineItem->getRequirement();
+
+        // if we don't have any requirement, then it's obviously valid
+        if (!$requirement) {
             return true;
         }
 
-        $scopeWithoutLineItem = new CartRuleScope($calculated, $context);
+        // build needed scope for rule and fetch needed order
+        // for flow rules if available
+        $scopeWithoutLineItem = $this->buildRuleScope($requirement, $calculated, $context);
+
+        if (!$scopeWithoutLineItem) {
+            return false;
+        }
 
         // set our currently registered group builder in our cart data
         // to be able to use that one within our line item rule
         $data = $scopeWithoutLineItem->getCart()->getData();
         $data->set(LineItemGroupBuilder::class, $this->groupBuilder);
 
-        return $lineItem->getRequirement()->match($scopeWithoutLineItem);
+        return $requirement->match($scopeWithoutLineItem);
+    }
+
+    /**
+     * Builds the rule scope for the given line item.
+     */
+    private function buildRuleScope(Rule $rule, Cart $cart, SalesChannelContext $context): CartRuleScope|FlowRuleScope|null
+    {
+        if (!$this->hasFlowRuleScope($rule)) {
+            return new CartRuleScope($cart, $context);
+        }
+
+        $order = $this->fetchOrder($context->getContext(), $cart);
+
+        if (!$order) {
+            return null;
+        }
+
+        return new FlowRuleScope($order, $cart, $context);
+    }
+
+    /**
+     * Checks if the given rule has a flow rule scope.
+     */
+    private function hasFlowRuleScope(Rule $rule): bool
+    {
+        if ($rule instanceof FlowRule) {
+            return true;
+        }
+
+        if ($rule instanceof Container) {
+            foreach ($this->flattenRule($rule) as $item) {
+                if ($item instanceof FlowRule) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Flattens the nested rules in the given container.
+     *
+     * @return Rule[]
+     */
+    private function flattenRule(Container $container): array
```

Comment:
if this function does not have any other use case besides `hasFlowRuleScope` I would prefer making `hasFlowRuleScope` recursive:
 ```php
private function hasFlowRuleScope(...$rules): bool
{
    foreach ($rules as $item) {
        if ($item instanceof FlowRule) {
            return true;
        } else if ($item instanceof Container) {
            return $this->hasFlowRuleScope(...$item->getRules());
        }
    }
    
    return false;
}
```